### PR TITLE
Get columns from information_schema if not present

### PR DESCRIPTION
**Pull request by diPhantxm**
```
root@spqr_router_1_1:/go# psql -p 6432 -U user1 -d db1
psql (15.10 (Debian 15.10-0+deb12u1), server 13.18 (Ubuntu 13.18-1.pgdg20.04+1))
Type "help" for help.

db1=# create schema test;
CREATE SCHEMA
db1=# create table test.test(id int, name text, t timestamp);
CREATE TABLE
db1=# insert into test.test values(10, 'test', NULL);
INSERT 0 1
db1=# select * from test.test;
 id | name | t 
----+------+---
 10 | test | 
(1 row)



root@spqr_shard_1:/# psql -p 6432 -U user1 -d db1
psql (13.18 (Ubuntu 13.18-1.pgdg20.04+1))
Type "help" for help.

db1=# select * from test.test;
 id | name | t 
----+------+---
 10 | test | 
(1 row)



root@spqr_shard_2:/# psql -p 6432 -U user1 -d db1
psql (13.18 (Ubuntu 13.18-1.pgdg20.04+1))
Type "help" for help.

db1=# select * from test.test;
 id | name | t 
----+------+---
(0 rows)
```

### CODE REVIEW

File: router/relay/relay.go [hunk 1]
```diff
@@ -1803,3 +1816,40 @@ func (rst *RelayStateImpl) ProcessMessage(
...
 }
+
+func (rst *RelayStateImpl) loadColumns(schemaName, tableName string) ([]string, error) {
+	if rst.Client().Server() == nil {
+		if err := rst.RerouteToRandomRoute(); err != nil {
+			return nil, err
+		}
+		defer func() { _ = rst.Unroute(rst.activeShards) }()
+	}
+	err := rst.Client().Server().Send(&pgproto3.Query{String: fmt.Sprintf(`SELECT column_name
```
Comment (Denchick) [inline router/relay/relay.go:1827]: Possible SQL injection here

File: router/relay/relay.go [hunk 2]
```diff
@@ -1803,3 +1818,63 @@ func (rst *RelayStateImpl) ProcessMessage(
 func (rst *RelayStateImpl) ConnMgr() poolmgr.PoolMgr {
 	return rst.manager
 }
+
+func (rst *RelayStateImpl) loadColumns(schemaName, tableName string) ([]string, error) {
```
Comment (reshke) [inline router/relay/relay.go:1822]: We should definitely use connection different from user one. The design i have in mind is separate component (cache, syscache or maybe schemaCache) with its own pool. which is shared among all RelayState's.

File: router/relay/relay.go [hunk 3]
```diff
@@ -1803,3 +1818,63 @@ func (rst *RelayStateImpl) ProcessMessage(
...
+		if err := rst.RerouteToRandomRoute(); err != nil {
+			return nil, err
+		}
+		defer func() { _ = rst.Unroute(rst.activeShards) }()
+	}
+
+	for _, msg := range []pgproto3.FrontendMessage{
+		&pgproto3.Sync{},
+		&pgproto3.Parse{Query: `SELECT column_name FROM information_schema.columns WHERE table_schema=$1 AND table_name=$2`},
+		&pgproto3.Describe{ObjectType: 'S'},
```
Comment (reshke) [inline router/relay/relay.go:1841]: We do not need Describe if we know resulting schema (single text column)

File: test/regress/tests/router/sql/shard_routing.sql [hunk 4]
```diff
@@ -65,7 +65,7 @@ SELECT * FROM xxtt1 a WHERE a.w_id = 21 and j + i != 0;
 SELECT * FROM xxtt1 a WHERE a.w_id = 21 and w_id <= 30 and j + i != 0;
 
 -- check that `INSERT FROM SELECT` works
```
Comment (reshke) [inline test/regress/tests/router/sql/shard_routing.sql:67]: I don't understand this change. if `INSERT INTO xx SELECT * FROM xx a WHERE a.w_id = 20;` not works its a regression.

File: router/cache/schema_cache.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	conn, err := c.pool.ConnectionHost(rand.Uint(), kr.ShardKey{Name: shardName, RW: false}, host)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		_ = c.pool.Discard(conn)
+	}()
+
+	for _, msg := range []pgproto3.FrontendMessage{
+		&pgproto3.Sync{},
```
Comment (reshke) [inline router/cache/schema_cache.go:62]: We dont need this sync, are we?

File: router/cache/schema_cache.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		preferAZ = config.RouterConfig().AvailabilityZone
+	}
+	return &SchemaCache{
+		shardMapping:      shardMapping,
+		pool:              pool.NewDBPool(shardMapping, &startup.StartupParams{}, preferAZ),
+		tableColumnsCache: sync.Map{},
+	}
+}
+
+func (c *SchemaCache) GetColumns(be *config.BackendRule, schemaName, tableName string) ([]string, error) {
```
Comment (reshke) [inline router/cache/schema_cache.go:34]: GetColumns should use its own berule, defined in config. We can also default this berule to first defined berule, if not configured. I not sure what are consequences of resetting  berule here every time.

**Review by reshke — APPROVED** [review]

**Pull request merged**